qemu/tests/win_virtio_driver_update_by_installer.py

```diff
@@ -0,0 +1,204 @@
+                                                            device_hwid)
+        error_context.context("Install 'qemu-guest-agent' package "
+                              "in guest.", logging.info)
+        s_inst, o_inst = session.cmd_status_output(gagent_install_cmd)
+        if s_inst != 0:
+            test.fail("qemu-guest-agent install failed,"
+                      " the detailed info:\n%s." % o_inst)
+
+    error_context.context("Upgrade virtio driver to original",
+                          logging.info)
+    change_virtio_media(params["cdrom_virtio"])
+    vm.send_key('meta_l-d')
+    time.sleep(30)
+    win_driver_utils.copy_file_to_samepath(session, test, params)
+    win_driver_utils.install_driver_by_installer(session, test,
+                                                 run_install_cmd,
+                                                 installer_pkg_check_cmd)
+    session = vm.reboot(session)  # workaround as Bug 1847856
+
+    error_context.context("Check if gagent version is correct.",
+                          logging.info)
+    actual_gagent_version = session.cmd_output(gagent_pkg_info_cmd).split()[-2]
+    if actual_gagent_version != expected_gagent_version:
+        test.fail("gagent version is not right, expected is %s but got %s"
+                  % (expected_gagent_version, actual_gagent_version))
+
+    wrong_ver_driver = []
+    not_signed_driver = []
+    for driver_name, device_name in zip(driver_name_list, device_name_list):
+        error_context.context("%s Driver Check" % driver_name, logging.info)
+        inf_path = win_driver_utils.get_driver_inf_path(session, test,
+                                                        media_type,
+                                                        driver_name)
+        expected_ver = session.cmd("type %s | findstr /i /r DriverVer.*=" %
+                                   inf_path, timeout=360)
+        expected_ver = expected_ver.strip().split(",", 1)[-1]
+        if not expected_ver:
+            test.error("Failed to find driver version from inf file")
+        if driver_name != "qemufwcfg":
+            logging.info("Target version is '%s'", expected_ver)
+            ver_list = win_driver_utils._pnpdrv_info(session, device_name,
+                                                     ["DriverVersion"])
+            if expected_ver not in ver_list:
+                wrong_ver_driver.append(driver_name)
+        chk_cmd = params["vio_driver_chk_cmd"] % device_name[0:30]
+        chk_output = session.cmd_output(chk_cmd, timeout=chk_timeout)
+        if "FALSE" in chk_output:
+            not_signed_driver.append(driver_name)
+        elif "TRUE" not in chk_output:
+            test.error("Device %s is not found in guest" % device_name)
+    if wrong_ver_driver:
+        test.fail("%s not the expected driver version" % wrong_ver_driver)
+    if not_signed_driver:
+        test.fail("%s not digitally signed!" % not_signed_driver)
```

Comment:
This piece of code seems totally same with the above script, right?

Reply:
yes，so I had better make it a function and put it to provider， right?  The reason why I not make a funtion is that it should only fit this installer  case.